File: add_mirrored_positions.py
class AddMirroredPositions:
    def __init__(self, SA, param, update_shot_list, player_function):
        self.SA = SA
        self.param = param
        self.update_shot_list = update_shot_list
        self.player_function = player_function

    def execute(self):
        print("Start mirroring ...")

        SAnew = {
            'Shot': [],
            'Table': [],
            'ShotIDsVisible': [],
            'ColumnsVisible': self.SA['ColumnsVisible'],
            'Current_si': self.SA['Current_si'],
            'Current_ti': self.SA['Current_ti'],
            'fullfilename': self.SA['fullfilename']
        }

        for si, shot in enumerate(self.SA['Shot']):
            if self.SA['Table']['Mirrored'][si] != 0:
                # Copy existing mirrored shot
                SAnew['Shot'].append(shot)
                SAnew['Table'].append(self.SA['Table'][si])
                SAnew['ShotIDsVisible'].append(self.SA['ShotIDsVisible'][si])
            else:
                # Create mirrored shots
                for i in range(1, 5):
                    mirrored_shot = self.mirror_shot(shot, i)
                    SAnew['Shot'].append(mirrored_shot)
                    mirrored_table_entry = self.SA['Table'][si].copy()
                    mirrored_table_entry['Mirrored'] = i
                    SAnew['Table'].append(mirrored_table_entry)
                    SAnew['ShotIDsVisible'].append(
                        self.SA['Table']['ShotID'][si] + i / 10
                    )

        # Overwrite SA with new data
        self.SA.update(SAnew)

        # Update the GUI
        self.update_shot_list()

        # Plot selected things
        self.player_function('plotcurrent')

        print(f"Mirroring done. Database has now {len(self.SA['Shot'])} shots")

    def mirror_shot(self, shot, mode):
        mirrored_shot = shot.copy()
        for bi in range(3):
            if mode in [2, 4]:
                # Mirror X-axis
                mirrored_shot['Route0'][bi]['x'] = self.param['size'][1] - shot['Route0'][bi]['x']
                mirrored_shot['Route'][bi]['x'] = self.param['size'][1] - shot['Route'][bi]['x']
            if mode in [3, 4]:
                # Mirror Y-axis
                mirrored_shot['Route0'][bi]['y'] = self.param['size'][0] - shot['Route0'][bi]['y']
                mirrored_shot['Route'][bi]['y'] = self.param['size'][0] - shot['Route'][bi]['y']
        return mirrored_shot

File: AddMirroredPositions.m
function AddMirroredPositions(~,~)
global SA param

disp(['Start mirroring ...'])
% What do do here?
% Mirror X
% Mirror Y
% Miroor X & Y
% Add for column to document the mirror action
hsl = findobj('Tag', 'uitable_shotlist');
sz = size(hsl.Data);

% Add SAnew.Shot(si)
% Add SAnew.Table(si)
% Add SAnew.ShotIDsVisible(si)
% Copy SAnew.ColumnsVisible
% Copy SAnew.Current_si
% Copy SAnew.Current_ShotID
% Copy SAnew.Current_ti
% Copy SAnew.fullfilename

% Inititalize shot counter for SAnew 
sinew = 0;

col_si = strcmp('ShotID', hsl.ColumnName);
col_mi = strcmp('Mirrored', hsl.ColumnName);

% Loop over all shots in the list
for ti = 1:sz(1)
    
    ShotID = cell2mat(hsl.Data(ti, col_si));
    mirrored = cell2mat(hsl.Data(ti, col_mi));
    
    [~, si, ~] = intersect(cell2mat(SA.Table.ShotID)+cell2mat(SA.Table.Mirrored)/10, ShotID+mirrored/10);
    Shot_Mirror_ID = ShotID + mirrored/10;
    
    if SA.Table.Mirrored{si} ~= 0
        % Shots are already mirrored, just copy the shot
            sinew = sinew + 1;

            % copy whole line in the Table
            SAnew.Table(sinew,:) = SA.Table(si,:);
            SAnew.ShotIDsVisible(sinew,1) = SA.ShotIDsVisible(si);

    else        
        % mirror Shots 
        
        for i = 1:4
            sinew = sinew + 1;

            % copy whole line in the Table
            SAnew.Table(sinew,:) = SA.Table(si,:);
            
            if i == 1
                % copy original data
                SAnew.Shot(sinew) = SA.Shot(si);
                SAnew.Table.Mirrored{sinew} = 1;
                
            elseif i == 2
                % Mirror LONG cushion values
                for bi = 1:3
                    SAnew.Shot(sinew).Route0(bi).t = SA.Shot(si).Route0(bi).t;
                    SAnew.Shot(sinew).Route(bi).t = SA.Shot(si).Route(bi).t;
                    SAnew.Shot(sinew).Route0(bi).x = param.size(2) - SA.Shot(si).Route0(bi).x;
                    SAnew.Shot(sinew).Route(bi).x = param.size(2) - SA.Shot(si).Route(bi).x;
                    SAnew.Shot(sinew).Route0(bi).y = SA.Shot(si).Route0(bi).y;
                    SAnew.Shot(sinew).Route(bi).y = SA.Shot(si).Route(bi).y;
                end
                SAnew.Table.Mirrored{sinew} = 2;
                
            elseif i == 3
                % Mirror SHORT cushion values
                for bi = 1:3
                    SAnew.Shot(sinew).Route0(bi).t = SA.Shot(si).Route0(bi).t;
                    SAnew.Shot(sinew).Route(bi).t = SA.Shot(si).Route(bi).t;
                    SAnew.Shot(sinew).Route0(bi).x = SA.Shot(si).Route0(bi).x;
                    SAnew.Shot(sinew).Route(bi).x = SA.Shot(si).Route(bi).x;
                    SAnew.Shot(sinew).Route0(bi).y = param.size(1) - SA.Shot(si).Route0(bi).y;
                    SAnew.Shot(sinew).Route(bi).y = param.size(1) - SA.Shot(si).Route(bi).y;
                end
                SAnew.Table.Mirrored{sinew} = 3;

            elseif i == 4
                % Mirror SHORT & LONG cushion values
                for bi = 1:3
                    SAnew.Shot(sinew).Route0(bi).t = SA.Shot(si).Route0(bi).t;
                    SAnew.Shot(sinew).Route(bi).t = SA.Shot(si).Route(bi).t;
                    SAnew.Shot(sinew).Route0(bi).x = param.size(2) - SA.Shot(si).Route0(bi).x;
                    SAnew.Shot(sinew).Route(bi).x = param.size(2) - SA.Shot(si).Route(bi).x;
                    SAnew.Shot(sinew).Route0(bi).y = param.size(1) - SA.Shot(si).Route0(bi).y;
                    SAnew.Shot(sinew).Route(bi).y = param.size(1) - SA.Shot(si).Route(bi).y;
                end
                SAnew.Table.Mirrored{sinew} = 4;
                
            end

            SAnew.ShotIDsVisible{sinew,1} = ShotID + SAnew.Table.Mirrored{sinew}/10;

        end

    end
end

% save old data to new
SAnew.ColumnsVisible = SA.ColumnsVisible;
SAnew.Current_si = SA.Current_si;
SAnew.Current_ti = SA.Current_ti;
SAnew.fullfilename = SA.fullfilename;

% Overwrite SA
 SA = SAnew;

% Update the GUI
update_ShotList

% Plot selected things
player.uptodate = 0;
PlayerFunction('plotcurrent',[])

%% done message
disp(['Mirroring done. Database has now ', num2str(length(SA.Shot)),' shots'])
